Multilanguage: deleting an associated item does not delete it in the _associations table

infograf768 opened the pull request:
See https://github.com/joomla/joomla-cms/issues/7127

As title says.
The _associations row should be deleted for the item as well as for the associated item if we have only 2 associated items.

To test:
Create a multilingual site and set associations to Yes in the Languagefilter plugin.

Create an article for each language (first with 2 languages) and associate them.
Note their id.
Send to trash. Don't empty trash yet.

With phpMyadmin look at the _associations table for rows with the id and the context "com_content.item"

Empty trash.
Reload the phpmyAdmin page: the 2 items are still there.

Recreate 2 articles and Patch.
re-test. Both rows are now deleted from the _associations table.

Then test with 3 languages, associating 3 articles.
Now, only the article deleted will be deleted from the associations table, letting the 2 others associated as should.

This is also working for menu items, categories, contacts, newsfeeds.

infograf768 committed:
Multilanguage: deleting an associated item does not delete it in the
_associations table

infograf768 commented:
This PR does not purge an existing _associations table from existing orphans associations.

infograf768 committed:
cs

Bakual commented on libraries/legacy/model/admin.php:
```diff
@@ -772,6 +772,53 @@ public function delete(&$pks)
+						$db->setQuery($query);
+						$key = $db->loadResult();
+
+						if (!empty($key))
```
No real need to `!empty($key)` here. Just using `if ($key)` does exactly the same thing.

Bakual commented:
I think it should be possible to get the count and key with one query instead of two using a join.
````
SELECT count(*), as1.key
FROM `k2wod_associations` as as1
LEFT jOIN `k2wod_associations` as as2 ON as1.key = as2.key
WHERE as1.context = "com_content.item"
AND as1.id = 3
````

infograf768 commented:
Please propose a PR against mine as I am not familiar with Join

Commit:
Optimise Queries

Merging the two queries to get the key and count into one.
I have merged the two queries to delete the entries into one as well. The only real difference between the two is the added WHERE clause for the id.

Bakual commented:
See https://github.com/infograf768/joomla-cms/pull/19

Another thing I noted: You're hardcoding the context as $this->option . '.item'. This makes this code useless for any extension which has more than one item type. Eg my own extension where I have com_sermonspeaker.sermon, com_sermonspeaker.speaker and com_sermonspeaker.serie.

I haven't found yet where that context is created when saving the item, but we should use the same code that is used there to determine the correct context :smile:

infograf768 commented:
>Another thing I noted: You're hardcoding the context as $this->option . '.item'

As far as I could see, all associations have that kind of context (look at the table). I tested with all core ones.
Will look where it is created

infograf768 commented:
In fact the context is created in each component when code exist for associations. It is not fetched from any existing code. It is arbitrary defined.
Example for article:

```
$query->clear()
						->insert('#__associations');

					foreach ($associations as $id)
					{
						$query->values($id . ',' . $db->quote('com_content.item') . ',' . $db->quote($key));
					}
```

Bakual commented:
Aww, crap. That should probably have used the available $context at the time when that code was written instead of creating an own one :(
So we basically have no clue if that context is true or not for any given extension.

So this code does only work for the extensions following that naming. For all others it will fail and maybe even create issues. Eg imagine an extension that has two contexts, one with .item and another with another suffix. Then this code could possibly delete the wrong asssociations.
I think best would be if we can specify the association context in the model as a property. If that is defined and Associations are enabled, then act. Otherwise do nothing.

This would mean that the code by default would not do anything if the extension didn't opt-in by defining the proper context.

Such a property could actually also be used in the saving part. I think the code there is always the same, with only the context changing in each model.

infograf768 committed:
Merge pull request #19 from Bakual/patch-2

Optimise Queries

infograf768 commented:
Thanks for the patch! It reduces the number of lines drastically.


I guess the reason of this context issue is that it was copied from the original associations which were only set for menu "item" in 2.5. When implementing the associations for other components, it was a sort of copy/paste.
As long as we are B/C by changing as you suggest (in another PR I guess), I am OK with it.
The way it is done is almost as if it was set in the model for core components.

I guess we would add to this delete functionality a call for the real defined context.
Would be good to get feedback from developers who have added multilingual associations to their components...

Bakual commented:
I need to think a bit and find time to do the PR then. If I forget, please poke me :smile:

infograf768 commented:
we could do something like:

lets add for the core components (here for article model)

```
/**
	 * The associations context for this content type (for example, 'com_content.item').
	 *
	 * @var      string
	 * @since    3.5
	 */
	public $associationsContext = 'com_content.item;
```
and then, in this PR, change:

`$assoc_context = $this->option . '.item';`

to

`$assoc_context = $this->associationsContext ? $this->associationsContext : $this->option . '.item';`

What do you think?

infograf768 commented:
We could even implement it in this PR if judged B/C.

Commit:
Add $associationsContext property and use that for the cleanup.

infograf768 commented:
Thinking loudly: we anyway will be B/C as the _associations table was never touched when deleting an item... Therefore we can implement the associationsContext and get it known by developers in 3.5.

infograf768 committed:
Merge pull request #20 from Bakual/assoc_deletion

Adding $associationsContext property to models

infograf768 commented:
Thanks @Bakual 
I am also in favour of your proposed simplification for the Save associations using $associationsContext in the components that use the JModelAdmin

Bakual commented:
You want it in this PR or as a follow up once this is merged?

infograf768 commented:
Your choice. We could consider it is a different topic as it is a small refactoring to simplify code (and help 3pd implement multilang in their extensions). So may be better to make it another PR, this one being merged first.

We need first another tester for this one. :smiley:

Bakual commented:
Agreed, it's quite a different topic and different to test.
Lets get that one merged first and afterwards do next :+1:

dgt41 commented:
tested ok with menu, category and article

infograf768 commented:
Thanks for testing. RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7583">issues.joomla.org/joomla-cms/7583</a>.</sub>

Kubik-Rubik commented:
Thank you @infograf768! Merged with https://github.com/joomla/joomla-cms/commit/76cef743361e6dc5be896af99ce8ffa927c1a3f2

infograf768 commented:
And thanks to @Bakual ;)
